src/client.js

```diff
@@ -0,0 +1,275 @@
+
+import Runner from './runner';
+import InterchangeFormat from './interchange-format';
+import {Joi, schemas, runSchema, MB, GB, TB} from './schemas';
+
+const MAX_S3_CHUNKS = 10000;
+
+/**
+ * This class represents a client of a service implemented using the
+ * `Controller` class.
+ */
+class Client {
+
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      runner: Joi.any(),
+      runnerOpts: Joi.object().default({}),
+      partsize: Joi.number().min(5*MB).max(5*GB).default(25*MB),
+      multisize: Joi.number().default(100*MB),
+    }).without('runner', 'runnerOpts')
+      .optionalKeys(['runner', 'runnerOpts']));
+
+    let {runner, runnerOpts, partsize, multisize} = opts;
+
+    // Store value and not reference
+    this.partsize = partsize+0;
+
+    // Unlike the Controller, which has much simplier usage, here we do want
+    // the full Runner api to be available to the client
+    if (!runner) {
+      runner = new Runner(opts.runnerOpts);
+    }
+    
+    // The Runner to use for this
+    this.runner = runner;
+
+    // The minimum size before switching to multipart
+    this.multisize = opts.multisize;
+  }
+
+
+  async __prepareSinglepartUpload(opts) {
+    let {filename} = opts;
+    let filestats = await fs.stat(filename);
+    let sha256 = crypto.createHash('sha256');
+    let size = 0;
+    let stream = fs.createReadStream(filename, {start: 0});
+    return new Promise((resolve, reject) => {
+      stream.on('error', reject);
+
+      stream.on('data', data => {
+        sha256.update(data);
+        size += data.length;
+      });
+
+      stream.on('end', async () => {
+        let finishedstats = await fs.stat(filename);
+        if (size !== filestats.size) {
+          throw new Error('File has a different number of bytes than was hashed');
+        } else if (finishedstats.size !== filestats.size) {
+          reject(new Error('File changed size during preperation'));
+        } else if (finishedstats.mtime.getTime() !== filestats.mtime.getTime()){
+          reject(new Error('File was modified during preperation'));
+        } else if (finishedstats.ino !== filestats.ino){
+          reject(new Error('File has changed inodes'));
+        } else {
+          sha256 = sha256.digest('hex');
+          resolve({
+            filename,
+            sha256,
+            size,
+          });
+        }
+      });
+    });
+  }
+
+  async __prepareMultipartUpload(opts) {
+    opts = opts || {};
+    let {filename, partsize} = opts;
+    // Ensure we're copying the value and not changing it
+    partsize = (partsize || this.partsize);
+
+    let sha256 = crypto.createHash('sha256');
+    let filestats = await fs.stat(filename);
+    let size = 0; // the computed size, to check against result of stat();
+    let partcount = Math.ceil(filestats.size / partsize);
+    let parts = [];
+
+    if (partsize/size < 2) {
+      throw new Error('Multipart upload must have at least 2 parts');
```

Comment:
if we remove this and allow `partcount == 1` then we don't need the singlepart thing...